### ISSUE TITLE
IDISP004 should not be triggered for serviceProvider.GetRequiredService<T>() in nullable context

### ISSUE THREAD

**Issue by jeremyVignelles** [DotNetAnalyzers/IDisposableAnalyzers#210]
The purpose of the service provider is to create and dispose the instances it contains. Thus, this line should not trigger a IDISP004.

```cs
new MyObject(this._serviceProvider.GetRequiredService<ILoggerFactory>());
```

EDIT: It seems to only happen when in a nullability-aware context, see below

**Comment by JohanLarsson** [comment]
Are you sure it is `this._serviceProvider.GetRequiredService<ILoggerFactory>()` and not `new MyObject(...)` that is triggering the warning?

**Comment by jeremyVignelles** [comment]
I moved that to a new line like this:
```cs
var loggerFactory = this._serviceProvider.GetRequiredService<ILoggerFactory>();
```
The message is now "Dispose created"

**Comment by jeremyVignelles** [comment]
```

    public class TestClass
    {
        public TestClass(IServiceProvider sp)
        {
            var loggerFactory = sp.GetRequiredService<IDisposable>();
        }
    }
```
Triggers IDISP001 as well... Maybe I should use another repo for tests like this..

**Comment by JohanLarsson** [comment]
> Maybe I should use another repo for tests like this..

How do you mean?

**Comment by jeremyVignelles** [comment]
Could not reproduce either outside of my application, I guess it's a version conflict somewhere. Closing for now.

**Comment by jeremyVignelles** [comment]
I managed to reproduce, the issue only happen when `<Nullable>enable</Nullable>` is set on the project

See https://github.com/jeremyVignelles/repro-idispanalyzers/commit/a0d63e9449354511b2830827771443097d68c9db#diff-9dbd58148108742e0731215aa147f9d6R35

**Comment by fuocor** [comment]
This started happening with netcore.app 3.1.3 and it's all over the place. Any IDisposable or IAsyncDisposable is being flagged.

**Comment by jeremyVignelles** [comment]
@JohanLarsson : Can the "needs-repro" tag be removed here?

**Comment by JohanLarsson** [comment]
I cloned but am not seeing the warning on that line. There is an IDISP004 warning [here](https://github.com/jeremyVignelles/repro-idispanalyzers/blob/master/ReproIDisposableAnalyzers/Program.cs#L16) though.

**Comment by jeremyVignelles** [comment]
@JohanLarsson : Did you checkout the repro/210 branch?

**Comment by JohanLarsson** [comment]
Yep, not sure what I dumbed. Can you update the branch to latest version of the analyzer just to be sure?

**Comment by jeremyVignelles** [comment]
I just cloned the projet on my home computer and I cannot reproduce either. Maybe the configuration is different from my work computer or a VS update "fixed" the issue?

I thought I reported that other IDISP004 on asp.net core initialization, but I can't find it anywhere.
It is a correct usage, since Run() is actually using() the created instance.

**Comment by jeremyVignelles** [comment]
@fuocor : Do you still have any issue in your code with the new version? Do my project warn about something on your side?

**Comment by fuocor** [comment]
Still there...
```
        private void ConfigureClientServices(HostBuilderContext unused, IServiceCollection services)
        {
            static IPublicClientApplication GetPublicClientApplication(IServiceProvider services)
            {
                var configuration = services.GetRequiredService<IConfiguration>();
                var clientId = configuration[ClientDefaults.Authentication.ClientId];
                Guard.Throw.ArgumentNullException(string.IsNullOrEmpty(clientId), nameof(clientId));
                return PublicClientApplicationBuilder
                    .Create(configuration[ClientDefaults.Authentication.ClientId])
                    .WithAuthority(AzureCloudInstance.AzurePublic, "7b948053-cc61-434b-af8e-9de586f466aa")
                    .Build();
            }

            services
                .AddSingleton(this.ServerHost.Services.GetRequiredService<ITelemetryChannel>())
                .AddSingleton(this.ServerHost.Services.GetRequiredService<TelemetryClient>())
                .AddSingleton(GetPublicClientApplication)
                .AddSingleton<IClientApplicationBase>(s => s.GetRequiredService<IPublicClientApplication>())
                .AddSingleton<TelemetryProvider, ChorusUnitTestApplication>()
                .AddSingleton<IGeocode, Geocode>();
        }
```
the ```this.ServerHost.Services.GetRequiredService<ITelemetryChannel>()``` fails. along with several dozen others. 

:(

**Comment by JohanLarsson** [comment]
Thanks for the repro code, I'll try some more when I find time.

**Comment by jeremyVignelles** [comment]
I faced the issue at work, and reopened my repro project at the state it was when I first sent it.
The issue was still there.

I pulled the code and upgraded to the latest version of IDisposableAnalyzers.

Still getting a IDISP001 - Dispose created (or IDISP004 if I don't put it in a variable.

There seems to be a difference between my home and my work installations of Visual Studio 2019. Any idea of what could cause such difference?

Work configuration:
- VS professional 2019 (english)
- Resharper 2020.1
- dotnet sdk 3.1.201

I still have some old SDK and a VS2017 installed on my computer, I don't know if that can cause such issues.

**Comment by JohanLarsson** [comment]
I would be surprised if ReSharper has anything to do with this, visual studio version is more likely. Unfortunate that the issue is present in the higher version if it is a regression there.

**Comment by jeremyVignelles** [comment]
Both version show me a warning, don't know how I couldn't reproduce the last time, and how you cannot reproduce. Did you try it again?

**Comment by JohanLarsson** [comment]
![image](https://user-images.githubusercontent.com/1640096/81212999-094c7380-8fd6-11ea-9259-9e931fec5b21.png)

I see it now in ValidCode.NetCore

**Comment by JohanLarsson** [comment]
https://www.nuget.org/packages/IDisposableAnalyzers/3.3.8
